src/lsm/k_way_merge.zig

```diff
@@ -24,207 +29,260 @@ pub fn KWayMergeIteratorType(
+                if (a.sentinel) return false;
+
+                const ordered = if (direction == .ascending) a.key < b.key else a.key > b.key;
+                const stabler = (a.key == b.key) and
+                    stream_precedence(ctx, a.stream_id, b.stream_id);
+                return ordered or stabler; // “true”  means  a  loses
+            }
+        };
 
         pub fn init(
             context: *Context,
-            streams_count: u32,
+            streams_count: u16,
             direction: Direction,
         ) KWayMergeIterator {
-            assert(streams_count <= streams_max);
+            assert(streams_count <= options.streams_max);
             // Streams ZERO can be used to represent empty sets.
             maybe(streams_count == 0);
 
             return .{
                 .context = context,
-                .streams_count = streams_count,
+                .tree_height = 0,
+                .nodes_count = 0,
+                .contender = sentinel,
+                .losers = .{sentinel} ** node_max,
+                .key_popped = null,
                 .direction = direction,
+                .streams_active = undefined,
+                .streams_count = streams_count,
                 .state = .loading,
             };
         }
 
-        pub fn reset(it: *KWayMergeIterator) void {
-            it.* = .{
-                .context = it.context,
-                .streams_count = it.streams_count,
-                .direction = it.direction,
+        pub fn reset(self: *KWayMergeIterator) void {
+            self.* = .{
+                .context = self.context,
+                .tree_height = 0,
+                .nodes_count = 0,
+                .contender = sentinel,
+                .losers = .{sentinel} ** node_max,
+                .direction = self.direction,
+                .streams_active = self.streams_active,
+                .streams_count = self.streams_count,
                 .state = .loading,
-                .key_popped = it.key_popped,
+                .key_popped = self.key_popped,
             };
         }
 
-        fn load(it: *KWayMergeIterator) error{Drained}!void {
-            assert(it.state == .loading);
-            assert(it.k == 0);
-
-            errdefer it.reset();
+        fn loading(
+            self: *KWayMergeIterator,
+        ) error{Drained}!void {
+            errdefer self.reset();
 
-            // We must loop on stream_index but assign at it.k, as k may be less than stream_index
-            // when there are empty streams.
-            var stream_index: u32 = 0;
-            while (stream_index < it.streams_count) : (stream_index += 1) {
-                it.keys[it.k] = stream_peek(it.context, stream_index) catch |err| switch (err) {
-                    error.Drained => return error.Drained,
+            // Collect the non‑empty batches as initial “contestants”.
+            var contestants: [node_max]Node = .{sentinel} ** node_max;
```

Comment:
Use `@splat` instead.
Also on `init()` and `reset()`.

```suggestion
            var contestants: [node_max]Node = @splat(sentinel);
```